File: src/algo/sort/05.mergeSort.js
function mergeSort(arr) {
    var len = arr.length;
    if(len < 2) {
        return arr;
    }

    var middle = Math.floor(len/2), 
        left = arr.slice(0,middle), 
        right = arr.slice(middle);
        
    return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right) {
    var result = [];

    while(left.length && right.length) {
        if(left[0] <= right[0]) {
            result.push(left.shift());
        } else {
            result.push(right.shift());
        }
    }

    while(left.length) {
        result.push(left.shift());
    }

    while(right.length) {
        result.push(right.shift());
    }

    return result;
}

File: src/algo/sort/05.mergeSort.py
def mergeSort(arr):
    import math

    if(len(arr) < 2):
        return arr
    
    middle = math.floor(len(arr)/2)
    left, right = arr[0:middle], arr[middle:]
    return merge(mergeSort(left), mergeSort(right))

def merge(left, right):
    result = []
    while left and right:
        if left[0] <= right[0]:
            result.append(left.pop(0))
        else:
            result.append(right.pop(0))
    
    while left:
        result.append(left.pop(0))
    while right:
        result.append(right.pop(0))

    return result


arr = [1,32,9,100,123,7,-1,0]
result = mergeSort(arr)
print(result)